tests/unit_tests/base32.cpp

```diff
@@ -0,0 +1,284 @@
+
+        EXPECT_EQ(raw_buf, decoded_buf);
+    }
+}
+*/
+
+TEST(base32, future_modification_protection)
+{
+    const boost::filesystem::path test_file_path = unit_test::data_dir / "base32" / "future_modification_protection.txt";
+
+    // pairs of (hex encoding of random bytes, base32_monero encoding of random bytes)
+    std::vector<std::pair<std::string, std::string>> test_cases;
+
+    // read test cases from data file
+    std::ifstream ifs(test_file_path.string());
+    ASSERT_TRUE(ifs);
+    while (ifs)
+    {
+        std::string hex_enc;
+        ifs >> hex_enc;
+
+        if (hex_enc.empty())
+            break;
+
+        std::string base32_enc;
+        ifs >> base32_enc;
+
+        ASSERT_FALSE(base32_enc.empty()); // we shouldn't run out of data on this part
+
+        test_cases.push_back({hex_enc, base32_enc});
+    }
+
+    ASSERT_EQ(249, test_cases.size()); // there should be 249 test cases in the file
+
+    for (const auto& test_case : test_cases)
+    {
+        std::string raw_buf;
+        ASSERT_TRUE(epee::string_tools::parse_hexstr_to_binbuff(test_case.first, raw_buf));
+
+        // test that base32_encode(hex_decode(test_case.first)) == test_case.second
+        const std::string encoded_buf = base32::encode(raw_buf);
+        EXPECT_EQ(test_case.second, encoded_buf);
+
+        // test that base32_decode(test_case.second) == hex_decode(test_cast.first)
+        const std::string decoded_buf = base32::decode(test_case.second);
+        EXPECT_EQ(raw_buf, decoded_buf);
+    }
+}
+
+TEST(base32, right_append_affects_right_enc)
+{
+    // test that we can append randomness on the right side of the unencoded message and keep
+    // the same prefix. this property makes Jamtis address headers less annoying
+
+    for (size_t left_len = 0; left_len < 250; ++left_len)
+    {
+        const size_t expected_prefix_untouched = left_len / 8 * 5;
+
+        std::string left_buf;
+        left_buf.resize(left_len);
+        crypto::generate_random_bytes_not_thread_safe(left_buf.size(), &left_buf[0]);
+
+        const std::string left_encoded = base32::encode(left_buf);
+
+        for (size_t right_len = 1; right_len <= 16; ++right_len)
+        {
+            std::string combined_buf = left_buf;
+            combined_buf.resize(left_len + right_len);
+            crypto::generate_random_bytes_not_thread_safe(right_len, &combined_buf[left_len]);
+
+            const std::string combined_encoded = base32::encode(combined_buf);
+
+            const size_t prefix_sim = num_prefix_similar(left_encoded.data(), combined_encoded.data(), left_len);
+
+            EXPECT_GE(prefix_sim, expected_prefix_untouched);
+        }
+    }
+}
+
+TEST(base32, right_modify_affects_right_enc)
+{
+    // test that we can randomly modify on the right side of the unencoded message and keep
+    // the same prefix. this property makes Jamtis address headers less annoying
```

Comment:
I don't quite get what you test here, why it's relevant and what's the connection with Jamtis address headers. Can you please elaborate a bit? (Probably my problem regarding understanding, not any problem with the test itself).

Reply:
It wasn't a hard requirement, but making sure that changes on the right side of the binary only affect the right side of the encoded string makes Jamtis address encoding harder to mess up because you can encode the header + body together and know that you'll get the correct prefix.